ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Forging.hs

```diff
@@ -31,195 +30,182 @@ import           Ouroboros.Consensus.HardFork.Combinator.Ledger (Ticked (..))
 import           Ouroboros.Consensus.HardFork.Combinator.Mempool
 import           Ouroboros.Consensus.HardFork.Combinator.Protocol
 import qualified Ouroboros.Consensus.HardFork.Combinator.State as State
+import qualified Ouroboros.Consensus.HardFork.Combinator.Util.Match as Match
 
 -- | If we cannot forge, it's because the current era could not forge
 type HardForkCannotForge xs = OneEraCannotForge xs
 
 type instance CannotForge (HardForkBlock xs) = HardForkCannotForge xs
 
--- | When we can lead in an era, we also have a 'ForgeStateInfo' for that era
--- (through 'BlockForging'). This means that when the hard fork combinator can
--- be a leader, it must at least have one 'ForgeStateInfo'. Hence
--- 'PerEraForgeStateInfo' is defined to be @OptNP False@.
-type HardForkForgeStateInfo xs = PerEraForgeStateInfo xs
+-- | For each era in which we want to forge blocks, we have a 'BlockForging',
+-- and thus 'ForgeStateInfo'.
+type HardForkForgeStateInfo xs = OneEraForgeStateInfo xs
 
 type instance ForgeStateInfo (HardForkBlock xs) = HardForkForgeStateInfo xs
 
--- | We update the forge state of each era for which we have a 'BlockForging'.
--- Each of those eras can individually fail.
---
--- When one or more eras fail to update, we consider the update to be failed.
--- This means we won't forge a block. For example, when the Shelley forge
--- state failed to update, we won't even forge a block in the Byron era.
-type HardForkForgeStateUpdateError xs = PerEraForgeStateUpdateError xs
+-- | For each era in which we want to forge blocks, we have a 'BlockForging',
+-- and thus 'ForgeStateUpdateError'.
+type HardForkForgeStateUpdateError xs = OneEraForgeStateUpdateError xs
 
 type instance ForgeStateUpdateError (HardForkBlock xs) =
   HardForkForgeStateUpdateError xs
 
 hardForkBlockForging ::
      forall m xs. (CanHardFork xs, Monad m)
-  => OptNP 'False (BlockForging m) xs
+  => NS (BlockForging m) xs
   -> BlockForging m (HardForkBlock xs)
-hardForkBlockForging blockForgings =
+hardForkBlockForging blockForging =
     BlockForging {
-        canBeLeader      = hmap (WrapCanBeLeader . canBeLeader) blockForgings
-      , updateForgeState = hardForkUpdateForgeState             blockForgings
-      , checkCanForge    = hardForkCheckCanForge                blockForgings
-      , forgeBlock       = hardForkForgeBlock                   blockForgings
+        forgeLabel       = hcollapse $ hmap (K . forgeLabel) blockForging
+      , canBeLeader      = hardForkCanBeLeader               blockForging
+      , updateForgeState = hardForkUpdateForgeState          blockForging
+      , checkCanForge    = hardForkCheckCanForge             blockForging
+      , forgeBlock       = hardForkForgeBlock                blockForging
       }
 
+hardForkCanBeLeader ::
+     CanHardFork xs
+  => NS (BlockForging m) xs -> HardForkCanBeLeader xs
+hardForkCanBeLeader =
+    OneEraCanBeLeader . hmap (WrapCanBeLeader . canBeLeader)
+
 hardForkUpdateForgeState ::
```

Comment:
Add postcondition: the returned `ForgeStateUpdateInfo` is from the same era as the given `NS BlockForging`.